duktape/src/main/jni/JavaScriptObject.cpp

```diff
@@ -0,0 +1,178 @@
+#else
+  m_instance = duk_get_heapptr(m_context, -1);
+  if (m_instance == nullptr) {
+#endif
+    duk_pop_2(m_context);
+    throw std::invalid_argument("JavaScript global called " + m_name + " is not an object");
+  }
+
+  if (duk_has_prop_string(m_context, -1, WRAPPER_THIS_PROP_NAME)) {
+    duk_pop_2(m_context);
+    throw std::invalid_argument("JavaScript object " + m_name + " is already proxied");
+  }
+
+  // Make sure that the object has all of the methods we want.
+  jmethodID getName = nullptr;
+  const jsize numArgs = env->GetArrayLength(methods);
+  for (jsize i = 0; i < numArgs; ++i) {
+    auto method = env->GetObjectArrayElement(methods, i);
+    if (getName == nullptr) {
+      jclass methodClass = env->GetObjectClass(method);
+      getName = env->GetMethodID(methodClass, "getName", "()Ljava/lang/String;");
+    }
+
+    const JString methodName(env, static_cast<jstring>(env->CallObjectMethod(method, getName)));
+    if (!duk_get_prop_string(m_context, -1, methodName)) {
+      duk_pop_2(m_context);
+      throw std::runtime_error("JavaScript global " + m_name + " has no method called " +
+          methodName.str());
+    } else if (!duk_is_callable(m_context, -1)) {
+      duk_pop_3(m_context);
+      throw std::runtime_error("JavaScript property " + m_name + "." + methodName.str() +
+          " not callable");
+    }
+
+    // TODO: build a call wrapper that handles marshalling the arguments and return value.
+    //       Store it in this object.
+
+    // Pop the method property.
+    duk_pop(m_context);
+  }
+
+#ifndef NO_INSTANCE_VAR
+  // Keep track of any previously registered finalizer.
+  duk_get_finalizer(m_context, -1);
+  m_nextFinalizer = duk_is_c_function(m_context, -1)
+     ? duk_get_c_function(m_context, -1)
+     : nullptr;
+  duk_pop(m_context);
+  duk_push_c_function(m_context, JavaScriptObject::finalizer, 1);
+  duk_set_finalizer(m_context, -2);
+#endif
+
+  duk_push_pointer(m_context, this);
+  duk_put_prop_string(m_context, -2, WRAPPER_THIS_PROP_NAME);
+
+  // Pop the global and our instance.
+  duk_pop_2(m_context);
+}
+
+JavaScriptObject::~JavaScriptObject() {
+  if (m_instance) {
+    // The instance still exists - detach from it.
+    duk_push_heapptr(m_context, m_instance);
+    duk_del_prop_string(m_context, -1, WRAPPER_THIS_PROP_NAME);
+
+    // Reset to the object's previous finalizer.
+    duk_push_c_function(m_context, m_nextFinalizer, 1);
+    duk_set_finalizer(m_context, -2);
+    duk_pop(m_context);
+  }
+}
+
+jobject JavaScriptObject::call(JNIEnv* env, jobject method, jobjectArray args) {
+#ifdef NO_INSTANCE_VAR
+  duk_push_global_object(m_context);
+  if (!duk_get_prop_string(m_context, -1, m_name.c_str())) {
+    duk_pop(m_context);
+    queueNullPointerException(env, "JavaScript object called " + m_name + " was not found");
+    return nullptr;
+  }
+#else
+  if (m_instance == nullptr) {
+    queueNullPointerException(env, "JavaScript object " + m_name + " has been garbage collected");
```

Comment:
`m_instance` points at the object that we're proxying in the Duktape VM.  I've set up the `finalizer` method to have Duktape let us know when that object is being GC'd so we can null out this pointer.    This could happen if the JavaScript code deletes/replaces the proxied variable.

Use of the proxy after closing Duktape will be detected before calling `DuktapeContext::call` in duktape-jni.cpp.